File: packages/assets/xmodelsurf.go
package assets

import (
	"encoding/binary"
	"fmt"
	"os"
)

type (
	XModelSurf struct {
		Name     string
		Version  uint16
		Surfaces []XmodelSurfSurface
	}

	XmodelSurfVertex struct {
		Normal   Vec3
		Color    Color
		UV       UV
		Bone     uint16
		Position Vec3
		Weights  []XmodelSurfWeight
	}

	XmodelSurfWeight struct {
		Bone      uint16
		Influence float32
	}

	XmodelSurfSurface struct {
		Vertices  []XmodelSurfVertex
		Triangles []Triangle
	}
)

const (
	RIGGED = 65535
)

// Load
func (xs *XModelSurf) Load(filePath string, xmodelPart *XModelPart) error {
	f, err := os.Open(filePath)
	if err != nil {
		return errorLogAndReturn(err)
	}
	defer f.Close()

	xs.Name = fileNameWithoutExt(filePath)

	err = binary.Read(f, binary.LittleEndian, &xs.Version)
	if err != nil {
		return errorLogAndReturn(err)
	}

	switch xs.Version {
	case VERSION_COD1:
		err := xs.loadV14(f, xmodelPart)
		if err != nil {
			return errorLogAndReturn(err)
		}

		return nil
	case VERSION_COD2:
		err := xs.loadV20(f, xmodelPart)
		if err != nil {
			return errorLogAndReturn(err)
		}

		return nil
	case VERSION_COD4:
		err := xs.loadV25(f, xmodelPart)
		if err != nil {
			return errorLogAndReturn(err)
		}

		return nil
	default:
		return fmt.Errorf("invalid xmodelsurf version: %v", xs.Version)
	}
}

// loadV14
func (xs *XModelSurf) loadV14(f *os.File, xmodelPart *XModelPart) error {
	var surfaceCount uint16
	err := binary.Read(f, binary.LittleEndian, &surfaceCount)
	if err != nil {
		return errorLogAndReturn(err)
	}

	for i := 0; i < int(surfaceCount); i++ {
		var surfaceHeader struct {
			_              byte
			VertexCount    uint16
			TriangleCount  uint16
			_              [2]byte
			DefaultBoneIdx uint16
		}

		err := binary.Read(f, binary.LittleEndian, &surfaceHeader)
		if err != nil {
			return errorLogAndReturn(err)
		}

		defaultBoneIdx := uint16(0)
		if surfaceHeader.DefaultBoneIdx == RIGGED {
			err := readPadding(f, 4)
			if err != nil {
				return errorLogAndReturn(err)
			}
		} else {
			defaultBoneIdx = surfaceHeader.DefaultBoneIdx
		}

		triangles := []Triangle{}
		for {
			var idxCount byte
			err := binary.Read(f, binary.LittleEndian, &idxCount)
			if err != nil {
				return errorLogAndReturn(err)
			}

			var idx1, idx2, idx3 uint16

			err = binary.Read(f, binary.LittleEndian, &idx1)
			if err != nil {
				return errorLogAndReturn(err)
			}
			err = binary.Read(f, binary.LittleEndian, &idx2)
			if err != nil {
				return errorLogAndReturn(err)
			}
			err = binary.Read(f, binary.LittleEndian, &idx3)
			if err != nil {
				return errorLogAndReturn(err)
			}

			if idx1 != idx2 && idx1 != idx3 && idx2 != idx3 {
				triangles = append(triangles, Triangle{
					V1: idx1,
					V2: idx2,
					V3: idx3,
				})
			}

			v := 0
			j := 3
			for j < int(idxCount) {
				idx4 := idx3
				var idx5 uint16
				err = binary.Read(f, binary.LittleEndian, &idx5)
				if err != nil {
					return errorLogAndReturn(err)
				}

				if idx4 != idx2 && idx4 != idx5 && idx2 != idx5 {
					triangles = append(triangles, Triangle{
						V1: idx4,
						V2: idx2,
						V3: idx5,
					})
				}

				v = j + 1
				if v >= int(idxCount) {
					break
				}

				idx2 = idx5
				err = binary.Read(f, binary.LittleEndian, &idx3)
				if err != nil {
					return errorLogAndReturn(err)
				}

				if idx4 != idx2 && idx4 != idx3 && idx2 != idx3 {
					triangles = append(triangles, Triangle{
						V1: idx4,
						V2: idx2,
						V3: idx3,
					})
				}

				j = v + 1
			}

			if len(triangles) >= int(surfaceHeader.TriangleCount) {
				break
			}
		}

		boneWeightCounts := make([]uint16, surfaceHeader.VertexCount)
		vertices := []XmodelSurfVertex{}
		for k := 0; k < int(surfaceHeader.VertexCount); k++ {
			vertex := XmodelSurfVertex{}
			err := binary.Read(f, binary.LittleEndian, &vertex.Normal)
			if err != nil {
				return errorLogAndReturn(err)
			}

			err = binary.Read(f, binary.LittleEndian, &vertex.UV)
			if err != nil {
				return errorLogAndReturn(err)
			}
			// flip UV
			vertex.UV.V = 1 - vertex.UV.V

			weightCount := uint16(0)
			vertexBoneIdx := defaultBoneIdx

			if surfaceHeader.DefaultBoneIdx == RIGGED {
				err = binary.Read(f, binary.LittleEndian, &weightCount)
				if err != nil {
					return errorLogAndReturn(err)
				}
				err = binary.Read(f, binary.LittleEndian, &vertexBoneIdx)
				if err != nil {
					return errorLogAndReturn(err)
				}
			}

			err = binary.Read(f, binary.LittleEndian, &vertex.Position)
			if err != nil {
				return errorLogAndReturn(err)
			}

			if weightCount != 0 {
				err = readPadding(f, 4)
				if err != nil {
					return errorLogAndReturn(err)
				}
			}

			boneWeightCounts[k] = weightCount

			if xmodelPart != nil {
				xmodelPartBone := xmodelPart.Bones[vertexBoneIdx]

				vertex.Position = xmodelPartBone.WorldTransform.Rotation.transformVec(vertex.Position)
				vertex.Position = vertex.Position.add(xmodelPartBone.WorldTransform.Positon)

				vertex.Normal = xmodelPartBone.WorldTransform.Rotation.transformVec(vertex.Normal)
			}

			vertex.Color = Color{
				R: 1,
				G: 1,
				B: 1,
				A: 1,
			}
			vertex.Bone = vertexBoneIdx
			vertex.Weights = append(vertex.Weights, XmodelSurfWeight{
				Bone:      vertexBoneIdx,
				Influence: 1,
			})

			vertices = append(vertices, vertex)
		}

		for m := 0; m < int(surfaceHeader.VertexCount); m++ {
			for n := 0; n < int(boneWeightCounts[m]); n++ {
				var weightBoneIdx uint16
				err := binary.Read(f, binary.LittleEndian, &weightBoneIdx)
				if err != nil {
					return errorLogAndReturn(err)
				}

				err = readPadding(f, 12)
				if err != nil {
					return errorLogAndReturn(err)
				}

				var weightInfluence float32
				err = binary.Read(f, binary.LittleEndian, &weightInfluence)
				if err != nil {
					return errorLogAndReturn(err)
				}

				weightInfluence = weightInfluence / RIGGED

				vertices[m].Weights[0].Influence -= weightInfluence
				vertices[m].Weights = append(vertices[m].Weights, XmodelSurfWeight{
					Bone:      weightBoneIdx,
					Influence: weightInfluence,
				})
			}
		}

		xs.Surfaces = append(xs.Surfaces, XmodelSurfSurface{
			Vertices:  vertices,
			Triangles: triangles,
		})
	}

	return nil
}

// loadV20
func (xs *XModelSurf) loadV20(f *os.File, xmodelPart *XModelPart) error {
	var surfaceCount uint16
	err := binary.Read(f, binary.LittleEndian, &surfaceCount)
	if err != nil {
		return errorLogAndReturn(err)
	}
	for i := 0; i < int(surfaceCount); i++ {
		var surfaceHeader struct {
			_              byte
			VertexCount    uint16
			TriangleCount  uint16
			DefaultBoneIdx uint16
		}

		err := binary.Read(f, binary.LittleEndian, &surfaceHeader)
		if err != nil {
			return errorLogAndReturn(err)
		}

		defaultBoneIdx := uint16(0)
		if surfaceHeader.DefaultBoneIdx == RIGGED {
			err := readPadding(f, 2)
			if err != nil {
				return errorLogAndReturn(err)
			}
		} else {
			defaultBoneIdx = surfaceHeader.DefaultBoneIdx
		}

		vertices := []XmodelSurfVertex{}
		for j := 0; j < int(surfaceHeader.VertexCount); j++ {
			vertex := XmodelSurfVertex{}
			err := binary.Read(f, binary.LittleEndian, &vertex.Normal)
			if err != nil {
				return errorLogAndReturn(err)
			}

			var vertexColor [4]byte
			err = binary.Read(f, binary.LittleEndian, &vertexColor)
			if err != nil {
				return errorLogAndReturn(err)
			}

			vertex.Color = Color{
				R: float32(vertexColor[0]) / 255,
				G: float32(vertexColor[1]) / 255,
				B: float32(vertexColor[2]) / 255,
				A: float32(vertexColor[3]) / 255,
			}

			err = binary.Read(f, binary.LittleEndian, &vertex.UV)
			if err != nil {
				return errorLogAndReturn(err)
			}
			vertex.UV.V = 1 - vertex.UV.V

			err = readPadding(f, 24)
			if err != nil {
				return errorLogAndReturn(err)
			}

			weightCount := uint8(0)
			vertexBoneIdx := defaultBoneIdx

			if surfaceHeader.DefaultBoneIdx == RIGGED {
				err := binary.Read(f, binary.LittleEndian, &weightCount)
				if err != nil {
					return errorLogAndReturn(err)
				}

				err = binary.Read(f, binary.LittleEndian, &vertexBoneIdx)
				if err != nil {
					return errorLogAndReturn(err)
				}
			}

			err = binary.Read(f, binary.LittleEndian, &vertex.Position)
			if err != nil {
				return errorLogAndReturn(err)
			}

			vertex.Weights = append(vertex.Weights, XmodelSurfWeight{
				Bone:      vertexBoneIdx,
				Influence: 1,
			})

			if weightCount > 0 {
				err := readPadding(f, 1)
				if err != nil {
					return errorLogAndReturn(err)
				}

				for k := 0; k < int(weightCount); k++ {
					var weightBoneIdx uint16
					err := binary.Read(f, binary.LittleEndian, &weightBoneIdx)
					if err != nil {
						return errorLogAndReturn(err)
					}

					err = readPadding(f, 12)
					if err != nil {
						return errorLogAndReturn(err)
					}

					var weightInfluence uint16
					err = binary.Read(f, binary.LittleEndian, &weightInfluence)
					if err != nil {
						return errorLogAndReturn(err)
					}

					weightInfluenceFloat := float32(weightInfluence) / RIGGED

					vertex.Weights[0].Influence -= weightInfluenceFloat
					vertex.Weights = append(vertex.Weights, XmodelSurfWeight{
						Bone:      weightBoneIdx,
						Influence: weightInfluenceFloat,
					})
				}
			}

			if xmodelPart != nil {
				xmodelPartBone := xmodelPart.Bones[vertexBoneIdx]

				vertex.Position = xmodelPartBone.WorldTransform.Rotation.transformVec(vertex.Position)
				vertex.Position = vertex.Position.add(xmodelPartBone.WorldTransform.Positon)

				vertex.Normal = xmodelPartBone.WorldTransform.Rotation.transformVec(vertex.Normal)
			}

			vertex.Bone = vertexBoneIdx

			vertices = append(vertices, vertex)
		}

		triangles := []Triangle{}
		for l := 0; l < int(surfaceHeader.TriangleCount); l++ {
			var tri Triangle
			err := binary.Read(f, binary.LittleEndian, &tri)
			if err != nil {
				return errorLogAndReturn(err)
			}

			triangles = append(triangles, tri)
		}

		xs.Surfaces = append(xs.Surfaces, XmodelSurfSurface{
			Vertices:  vertices,
			Triangles: triangles,
		})
	}

	return nil
}

// loadV25
func (xs *XModelSurf) loadV25(f *os.File, xmodelPart *XModelPart) error {
	var surfaceCount uint16
	err := binary.Read(f, binary.LittleEndian, &surfaceCount)
	if err != nil {
		return errorLogAndReturn(err)
	}

	for i := 0; i < int(surfaceCount); i++ {
		var surfaceHeader struct {
			_             [3]byte
			VertexCount   uint16
			TriangleCount uint16
			VertexCount2  uint16
		}

		err := binary.Read(f, binary.LittleEndian, &surfaceHeader)
		if err != nil {
			return errorLogAndReturn(err)
		}

		if surfaceHeader.VertexCount != surfaceHeader.VertexCount2 {
			for {
				var p uint16
				err := binary.Read(f, binary.LittleEndian, &p)
				if err != nil {
					return errorLogAndReturn(err)
				}

				if p == 0 {
					break
				}
			}

			err := readPadding(f, 2)
			if err != nil {
				return errorLogAndReturn(err)
			}
		} else {
			err := readPadding(f, 4)
			if err != nil {
				return errorLogAndReturn(err)
			}
		}

		vertices := []XmodelSurfVertex{}
		for j := 0; j < int(surfaceHeader.VertexCount); j++ {
			vertex := XmodelSurfVertex{}

			err := binary.Read(f, binary.LittleEndian, &vertex.Normal)
			if err != nil {
				return errorLogAndReturn(err)
			}

			var clr [4]byte
			err = binary.Read(f, binary.LittleEndian, &clr)
			if err != nil {
				return errorLogAndReturn(err)
			}

			vertex.Color = Color{
				R: float32(clr[0]) / 255,
				G: float32(clr[1]) / 255,
				B: float32(clr[2]) / 255,
				A: float32(clr[3]) / 255,
			}

			err = binary.Read(f, binary.LittleEndian, &vertex.UV)
			if err != nil {
				return errorLogAndReturn(err)
			}

			vertex.UV.V = 1 - vertex.UV.V

			err = readPadding(f, 24)
			if err != nil {
				return errorLogAndReturn(err)
			}

			weightCount := uint8(0)
			vertexBoneIdx := uint16(0)
			if surfaceHeader.VertexCount != surfaceHeader.VertexCount2 {
				err := binary.Read(f, binary.LittleEndian, &weightCount)
				if err != nil {
					return errorLogAndReturn(err)
				}

				err = binary.Read(f, binary.LittleEndian, &vertexBoneIdx)
				if err != nil {
					return errorLogAndReturn(err)
				}
			}

			err = binary.Read(f, binary.LittleEndian, &vertex.Position)
			if err != nil {
				return errorLogAndReturn(err)
			}

			vertex.Weights = append(vertex.Weights, XmodelSurfWeight{
				Bone:      vertexBoneIdx,
				Influence: 1,
			})

			if weightCount > 0 {
				for k := 0; k < int(weightCount); k++ {
					var weightBoneIdx uint16
					err := binary.Read(f, binary.LittleEndian, &weightBoneIdx)
					if err != nil {
						return errorLogAndReturn(err)
					}

					var weightInfluence uint16
					err = binary.Read(f, binary.LittleEndian, &weightInfluence)
					if err != nil {
						return errorLogAndReturn(err)
					}

					weightInfluenceFloat := float32(weightInfluence) / RIGGED
					vertex.Weights[0].Influence -= weightInfluenceFloat
					vertex.Weights = append(vertex.Weights, XmodelSurfWeight{
						Bone:      weightBoneIdx,
						Influence: weightInfluenceFloat,
					})
				}
			}

			vertex.Bone = vertexBoneIdx
			vertices = append(vertices, vertex)
		}

		triangles := []Triangle{}
		for l := 0; l < int(surfaceHeader.TriangleCount); l++ {
			var tri Triangle
			err := binary.Read(f, binary.LittleEndian, &tri)
			if err != nil {
				return errorLogAndReturn(err)
			}
		}

		xs.Surfaces = append(xs.Surfaces, XmodelSurfSurface{
			Vertices:  vertices,
			Triangles: triangles,
		})
	}

	return nil
}

File: rust/cod_asset_importer/src/assets/xmodelsurf.rs
use super::xmodel::XModelVersion;
use super::xmodelpart::XModelPart;
use crate::utils::{
    binary,
    error::Error,
    math::{color_from_vec, uv_from_vec, vec3_add, vec3_from_vec, vec3_rotate, Color, Vec3, UV},
    path::file_name_without_ext,
    Result,
};
use std::{fs::File, path::PathBuf};

pub const ASSETPATH: &str = "xmodelsurfs";
const RIGGED: i32 = 65535;

pub struct XModelSurf {
    pub name: String,
    pub version: u16,
    pub surfaces: Vec<XModelSurfSurface>,
}

pub struct XModelSurfVertex {
    pub normal: Vec3,
    pub color: Color,
    pub uv: UV,
    pub bone: u16,
    pub position: Vec3,
    pub weights: Vec<XModelSurfWeight>,
}

pub struct XModelSurfWeight {
    pub bone: u16,
    pub influence: f32,
}

pub struct XModelSurfSurface {
    pub vertices: Vec<XModelSurfVertex>,
    pub triangles: Vec<u16>,
}

impl XModelSurf {
    pub fn load(file_path: PathBuf, xmodel_part: Option<XModelPart>) -> Result<XModelSurf> {
        let mut file = File::open(&file_path)?;
        let name = file_name_without_ext(file_path);
        let version = binary::read::<u16>(&mut file)?;
        let mut xmodel_surf = XModelSurf {
            name,
            version,
            surfaces: Vec::new(),
        };

        match XModelVersion::valid(version) {
            Some(XModelVersion::V14) => {
                xmodel_surf.load_v14(&mut file, xmodel_part)?;
                Ok(xmodel_surf)
            }
            Some(XModelVersion::V20) => {
                xmodel_surf.load_v20(&mut file, xmodel_part)?;
                Ok(xmodel_surf)
            }
            Some(XModelVersion::V25) => {
                xmodel_surf.load_v25(&mut file)?;
                Ok(xmodel_surf)
            }
            None => Err(Error::new(format!(
                "invalid xmodelsurf version {}",
                version
            ))),
        }
    }

    fn load_v14(&mut self, file: &mut File, xmodel_part: Option<XModelPart>) -> Result<()> {
        let surface_count = binary::read::<u16>(file)?;

        for _ in 0..surface_count {
            binary::skip(file, 1)?;

            let vertex_count = binary::read::<u16>(file)?;
            let triangle_count = binary::read::<u16>(file)?;

            binary::skip(file, 2)?;

            let og_default_bone_idx = binary::read::<u16>(file)?;
            let default_bone_idx = if og_default_bone_idx as i32 == RIGGED {
                binary::skip(file, 4)?;
                0
            } else {
                og_default_bone_idx
            };

            let mut triangles: Vec<u16> = Vec::new();
            loop {
                let idx_count = binary::read::<u8>(file)?;

                let idx1 = binary::read::<u16>(file)?;
                let mut idx2 = binary::read::<u16>(file)?;
                let mut idx3 = binary::read::<u16>(file)?;

                if idx1 != idx2 && idx1 != idx3 && idx2 != idx3 {
                    triangles.extend_from_slice(&[idx3, idx2, idx1]);
                }

                let mut i = 3;
                while i < idx_count {
                    let idx4 = idx3;
                    let idx5 = binary::read::<u16>(file)?;

                    if idx4 != idx2 && idx4 != idx5 && idx2 != idx5 {
                        triangles.extend_from_slice(&[idx5, idx2, idx4]);
                    }

                    let v = i + 1;
                    if v >= idx_count {
                        break;
                    }

                    idx2 = idx5;
                    idx3 = binary::read::<u16>(file)?;

                    if idx4 != idx2 && idx4 != idx3 && idx2 != idx3 {
                        triangles.extend_from_slice(&[idx3, idx2, idx4]);
                    }

                    i = v + 1;
                }

                if triangles.len() / 3 >= triangle_count as usize {
                    break;
                }
            }

            let mut bone_weight_counts = vec![0u16; vertex_count as usize];
            let mut vertices: Vec<XModelSurfVertex> = Vec::new();
            for i in 0..vertex_count {
                let normal = binary::read_vec::<f32>(file, 3)?;
                let mut normal = vec3_from_vec(normal).unwrap();

                let uv = binary::read_vec::<f32>(file, 2)?;
                let mut weight_count = 0;
                let mut vertex_bone_idx = default_bone_idx;

                if og_default_bone_idx as i32 == RIGGED {
                    weight_count = binary::read::<u16>(file)?;
                    vertex_bone_idx = binary::read::<u16>(file)?;
                }

                let position = binary::read_vec::<f32>(file, 3)?;
                let mut position = vec3_from_vec(position).unwrap();

                if weight_count != 0 {
                    binary::skip(file, 4)?;
                }

                bone_weight_counts[i as usize] = weight_count;

                if let Some(xmodel_part) = xmodel_part.to_owned() {
                    let xmodel_part_bone = xmodel_part.bones[vertex_bone_idx as usize].to_owned();

                    position = vec3_rotate(position, xmodel_part_bone.world_transform.rotation);
                    position = vec3_add(position, xmodel_part_bone.world_transform.position);
                    normal = vec3_rotate(normal, xmodel_part_bone.world_transform.rotation);
                }

                vertices.push(XModelSurfVertex {
                    normal,
                    color: [1.0, 1.0, 1.0, 1.0],
                    uv: uv_from_vec(uv, true).unwrap(),
                    bone: vertex_bone_idx,
                    position,
                    weights: vec![XModelSurfWeight {
                        bone: vertex_bone_idx,
                        influence: 1.0,
                    }],
                })
            }

            for i in 0..vertex_count {
                for _ in 0..bone_weight_counts[i as usize] {
                    let weight_bone_idx = binary::read::<u16>(file)?;

                    binary::skip(file, 12)?;

                    let mut weight_influence = binary::read::<f32>(file)?;
                    weight_influence /= RIGGED as f32;

                    vertices[i as usize].weights[0].influence -= weight_influence;
                    vertices[i as usize].weights.push(XModelSurfWeight {
                        bone: weight_bone_idx,
                        influence: weight_influence,
                    });
                }
            }

            self.surfaces.push(XModelSurfSurface {
                vertices,
                triangles,
            });
        }

        Ok(())
    }

    fn load_v20(&mut self, file: &mut File, xmodel_part: Option<XModelPart>) -> Result<()> {
        let surface_count = binary::read::<u16>(file)?;

        for _ in 0..surface_count {
            binary::skip(file, 1)?;

            let vertex_count = binary::read::<u16>(file)?;
            let triangle_count = binary::read::<u16>(file)?;
            let og_default_bone_idx = binary::read::<u16>(file)?;
            let default_bone_idx: u16 = if og_default_bone_idx as i32 == RIGGED {
                binary::skip(file, 2)?;
                0
            } else {
                og_default_bone_idx
            };

            let mut vertices: Vec<XModelSurfVertex> = Vec::new();
            for _ in 0..vertex_count {
                let normal = binary::read_vec::<f32>(file, 3)?;
                let mut normal = vec3_from_vec(normal).unwrap();

                let color = binary::read_vec::<u8>(file, 4)?;
                let uv = binary::read_vec::<f32>(file, 2)?;

                binary::skip(file, 24)?;

                let mut weight_count = 0;
                let mut vertex_bone_idx = default_bone_idx;

                if og_default_bone_idx as i32 == RIGGED {
                    weight_count = binary::read::<u8>(file)?;
                    vertex_bone_idx = binary::read::<u16>(file)?;
                }

                let position = binary::read_vec::<f32>(file, 3)?;
                let mut position = vec3_from_vec(position).unwrap();

                let mut vertex_weights: Vec<XModelSurfWeight> = vec![XModelSurfWeight {
                    bone: vertex_bone_idx,
                    influence: 1.0,
                }];

                if weight_count > 0 {
                    binary::skip(file, 1)?;

                    for _ in 0..weight_count {
                        let weight_bone_idx = binary::read::<u16>(file)?;
                        binary::skip(file, 12)?;
                        let weight_influence = binary::read::<u16>(file)?;
                        let weight_influence = weight_influence as f32 / RIGGED as f32;

                        vertex_weights[0].influence -= weight_influence;
                        vertex_weights.push(XModelSurfWeight {
                            bone: weight_bone_idx,
                            influence: weight_influence,
                        });
                    }
                }

                if let Some(xmodel_part) = xmodel_part.to_owned() {
                    let xmodel_part_bone = xmodel_part.bones[vertex_bone_idx as usize].to_owned();

                    position = vec3_rotate(position, xmodel_part_bone.world_transform.rotation);
                    position = vec3_add(position, xmodel_part_bone.world_transform.position);
                    normal = vec3_rotate(normal, xmodel_part_bone.world_transform.rotation);
                }

                vertices.push(XModelSurfVertex {
                    normal,
                    color: color_from_vec(color).unwrap(),
                    uv: uv_from_vec(uv, false).unwrap(),
                    bone: vertex_bone_idx,
                    position,
                    weights: vertex_weights,
                });
            }

            let mut triangles: Vec<u16> = Vec::new();
            for _ in 0..triangle_count {
                let t = binary::read_vec::<u16>(file, 3)?;
                triangles.extend_from_slice(&[t[0], t[2], t[1]]);
            }

            self.surfaces.push(XModelSurfSurface {
                vertices,
                triangles,
            });
        }

        Ok(())
    }

    fn load_v25(&mut self, file: &mut File) -> Result<()> {
        let surface_count = binary::read::<u16>(file)?;

        for _ in 0..surface_count {
            binary::skip(file, 3)?;
            let vertex_count = binary::read::<u16>(file)?;
            let triangle_count = binary::read::<u16>(file)?;
            let vertex_count2 = binary::read::<u16>(file)?;

            if vertex_count != vertex_count2 {
                binary::skip(file, 2)?;
                if vertex_count2 != 0 {
                    loop {
                        let p = binary::read::<u16>(file)?;
                        if p == 0 {
                            break;
                        }
                    }
                    binary::skip(file, 2)?;
                }
            } else {
                binary::skip(file, 4)?;
            }

            let mut vertices: Vec<XModelSurfVertex> = Vec::new();
            for _ in 0..vertex_count {
                let normal = binary::read_vec::<f32>(file, 3)?;
                let color = binary::read_vec::<u8>(file, 4)?;
                let uv = binary::read_vec::<f32>(file, 2)?;

                binary::skip(file, 24)?;

                let mut weight_count = 0;
                let mut vertex_bone_idx = 0;
                if vertex_count != vertex_count2 {
                    weight_count = binary::read::<u8>(file)?;
                    vertex_bone_idx = binary::read::<u16>(file)?;
                }

                let position = binary::read_vec::<f32>(file, 3)?;

                let mut vertex_weights: Vec<XModelSurfWeight> = vec![XModelSurfWeight {
                    bone: vertex_bone_idx,
                    influence: 1.0,
                }];

                if weight_count > 0 {
                    for _ in 0..weight_count {
                        let weight_bone_idx = binary::read::<u16>(file)?;
                        let weight_influence = binary::read::<u16>(file)?;
                        let weight_influence = weight_influence as f32 / RIGGED as f32;
                        vertex_weights[0].influence -= weight_influence;
                        vertex_weights.push(XModelSurfWeight {
                            bone: weight_bone_idx,
                            influence: weight_influence,
                        });
                    }
                }

                vertices.push(XModelSurfVertex {
                    normal: vec3_from_vec(normal).unwrap(),
                    color: color_from_vec(color).unwrap(),
                    uv: uv_from_vec(uv, false).unwrap(),
                    bone: vertex_bone_idx,
                    position: vec3_from_vec(position).unwrap(),
                    weights: vertex_weights,
                })
            }

            let mut triangles: Vec<u16> = Vec::new();
            for _ in 0..triangle_count {
                let t = binary::read_vec::<u16>(file, 3)?;
                triangles.extend_from_slice(&[t[0], t[2], t[1]]);
            }

            self.surfaces.push(XModelSurfSurface {
                vertices,
                triangles,
            });
        }

        Ok(())
    }
}
